c++/src/vector.cc

```diff
@@ -0,0 +1,100 @@
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <initializer_list>
+#include <cassert>
+
+template <class T>
+class Vector {
+private:
+  unsigned int length = 0;
+  unsigned int capacity = 8;
+  T* dataStore = nullptr;
+
+public:
+  Vector() : dataStore(new T[this->capacity]{0}) {}
+
+  Vector(std::initializer_list<T> list) 
+  : dataStore(new T[this->capacity]{0}) {
+    for (const T &item : list) {
+      this->push(item);
+    }
+  }
+
+  ~Vector() {
+    delete[] this->dataStore;
+  }
+
+
+  void push(T val) {
+    if (length >= capacity) {
+      std::cout << "Vector capacity exceeded. ";
+      std::exit(1);
+    }
+
+    this->dataStore[this->length] = val;
+    this->length++;
+  }
+
+  T pop() {
+    if (length == 0) {
+      std::cout << "Attempted to pop empty vector. ";
+      std::exit(1);
+    }
+
+    length--;
+    return this->dataStore[length + 1];
```

Comment:
This reads very weird. You subtract one but then you add one cause you wanted the previous value, but you want it to stay subtracted. This feels like a perfect opportunity for post-decrement, which gives you the current value for the place you're using it, but ensures that the value is decremented by the end of the statement.

``` cpp
return this->dataStore[this->length--];
```

You can think of that as expanding more or less to:

``` cpp
T val = this->dataStore[length];
length--;
return val;
```

Which I think would look better than what you have. The reason what you have is confusing (relative to the other options) is that you change length then you change it back for a temporary reason.